### PR TITLE
Implement anonymous user promotion

### PR DESCRIPTION
**Pull request by kiootic**
requires #1407 
ref #1376

### CODE REVIEW

File: pkg/auth/dependency/webapp/authenticate_provider_impl.go [hunk 1]
```diff
@@ -483,7 +492,7 @@ func (p *AuthenticateProviderImpl) LinkIdentityProvider(w http.ResponseWriter, r
 	q.Set("redirect_uri", r.URL.Path)
 	webappSSOState.SetRequestQuery(q.Encode())
 	state := sso.State{
-		Action: "link",
+		Action: action,
```
Comment (louischan-oursky) [inline pkg/auth/dependency/webapp/authenticate_provider_impl.go:495]: I think we need to fix the `userID` if it is promotion? The `userID` in this function is from `auth.GetSession`

File: pkg/auth/dependency/webapp/translation.go [hunk 2]
```diff
@@ -25,10 +25,15 @@ var TemplateAuthUITranslationJSON = template.Spec{
...
 	"sign-in-apple": "Sign in with Apple",
+	"sign-up-apple": "Sign up with Apple",
 	"sign-in-google": "Sign in with Google",
+	"sign-up-google": "Sign up with Google",
 	"sign-in-facebook": "Sign in with Facebook",
+	"sign-up-facebook": "Sign up with Facebook",
 	"sign-in-linkedin": "Sign in with LinkedIn",
+	"sign-up-linkedin": "Sign up with LinkedIn",
 	"sign-in-azureadv2": "Sign in with Azure AD",
+	"sign-up-azureadv2": "Sign up with Azure AD",
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/auth/dependency/webapp/translation.go:36]: It is common to have different text to be consistent with the page (e.g. sign up with XXX for sign up page, log in with XXX for log in page), e.g. [Stack Overflow](https://stackoverflow.com/users/signup).

File: pkg/auth/dependency/webapp/authenticate_provider_impl.go (hunk 1 above)
Comment (kiootic) [reply pkg/auth/dependency/webapp/authenticate_provider_impl.go:495]: Indeed, I didn't saw it when rebasing.

File: pkg/auth/dependency/webapp/translation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/auth/dependency/webapp/translation.go:36]: Fair enough.

File: pkg/auth/dependency/interaction/flows/webapp.go [hunk 3]
```diff
@@ -114,6 +114,52 @@ func (f *WebAppFlow) SignupWithLoginID(loginIDKey, loginID string) (*WebAppResul
...
+	// Trigger OOB OTP if the first authenticator is OOB OTP
+	if len(s.Steps[0].AvailableAuthenticators) > 0 && s.Steps[0].AvailableAuthenticators[0].Type == authn.AuthenticatorTypeOOB {
+		step = WebAppStepSetupOOBOTP
+		err = f.Interactions.PerformAction(i, interaction.StepSetupPrimaryAuthenticator, &interaction.ActionTriggerOOBAuthenticator{
+			Authenticator: s.Steps[0].AvailableAuthenticators[0],
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (louischan-oursky) [inline pkg/auth/dependency/interaction/flows/webapp.go:148]: Line 134 - Line 148 can adjusted according to https://github.com/SkygearIO/skygear-server/pull/1407/commits/ac9091249d2b38c63173156a91a92c008df99add and https://github.com/SkygearIO/skygear-server/pull/1407/commits/0362d54743d12dd7031712c3b150b46ca427901a

File: pkg/auth/dependency/interaction/flows/anonymous.go [hunk 4]
```diff
@@ -101,3 +101,18 @@ func (f *AnonymousFlow) Authenticate(requestJWT string, clientID string) (*authn
 
 	return attrs, nil
 }
+
+func (f *AnonymousFlow) DecodeUserID(requestJWT string) (string, anonymous.RequestAction, error) {
+	identity, request, err := f.Anonymous.ParseRequest(requestJWT)
```
Comment (carmenlau) [inline pkg/auth/dependency/interaction/flows/anonymous.go:106]: Seems `Anonymous.ParseRequest` may return nil identity if kid not exists and key is provided in jwt? Maybe we should check and return `ErrInvalidCredentials` if identity is nil here.

File: pkg/auth/dependency/interaction/flows/webapp.go (hunk 3 above)
Comment (louischan-oursky) [reply pkg/auth/dependency/interaction/flows/webapp.go:148]: Ping

**Pull request merged**